### ISSUE TITLE
Handle HTTP request timeouts nicely

### ISSUE THREAD

**Issue by benoit74** [openzim/zimfarm#832]
Issue #826 and others have shown that timeouts (code handling an HTTP request running longer than 1 minute) are not very properly handled in the dispatcher/backend API:
- sometimes we have an HTTP 499 error, sometimes an HTTP 504 error (looks mostly random at first sight)
- code is continuing to run, potential forever if stuck in a dead loop, and not releasing any resources (e.g. DB locks) ; it is not clear if it is always the case or only on some conditions

We should investigate:
- why we have various errors and if possible have some logic / reasoning around it
- how could we stop code running in case of code running for too long + return a "better" error (e.g. probably handle the timeout in the code instead of nginx/UWSGI middleware)

**Comment by rgaudin** [comment]
I thought we added timeouts (`timeout=`) to all the HTTP requests sent from the API. Is that right? If that's the case, then requests that last this long would simply raise an exception.

My point is, I don't think we should focus on HTTP requests if those have been handled. If we still see 499/504 and DB locks (not clear in your message), it might be that we improperly react to Exceptions and/or that we have other parts that runs for too long.

I think a good first step would be to document what's going on:

- k8s LB timeouts
- API nginx timeouts
- requests behavior (upgrading requests is probably a wise as well)
- Flask timeouts (if any)
- uwsgi timeouts (if any)
- DB lock conditions (when does the ORM sets and unsets them)
- requests that are subject to those 499/504
- do we know where those HTTP responses are raises? one of the nginx? flask? uswgi?

**Comment by benoit74** [comment]
Yes, we have added timeouts to all HTTP requests.
No, there is no more 499/504 or DB locks.

My point is more that our recent experience around sub-requests taking forever and broadcaster not returning are just two situations of a more general pattern: it is hard (impossible) to be sure that all code paths in all API endpoints will complete in due time.

And the current setup is returned an HTTP response 504/499 to the client, while code might still be running in the backend.

This is not a good setup because:
- it means that you could pile up resources consumption (e.g. DB locks) if the client retries too many times ; in the current situation the worker is not retrying much, so it limits the blast radius but it could be worse under different situations
- it means that you have code which is very difficult to debug because it might produce effects after the HTTP response has been sent (for instance if it is just a query which is taking minutes to complete but succeed in the end)

To solve this issue I think we should have a good cooperation between all the timeouts you mentioned. 

And to achieve this goal, I totally agree with you that we **first** need a better understanding of all of these timeouts and which component is raising these 504/499 responses.

Regarding DB lock, I can already provide an answer:
- the lock is set when the DB session is flushed
- the lock is released when the DB session is committed or rolled-back (i.e when we exit the session context manager with current code)

**Comment by rgaudin** [comment]
Very interesting. So whenever we use _preflush_ in order to call a different piece of code (using different session), we are locking very ahead of the release.

I have the notification case in mind where we flush before calling the notification code (which itself does long stuff like requests) and only once this finishes, do we commit.
Given what you said, I see:
- possible accumulation of locks if we have many similar incoming requests and delays in notification processing
- wonder how locks behaves when we flush (and lock) then other code updates same data with its own flush+lock+commit+unlock. Maybe we should share session instead
- not certain we're releasing everything should that sub-code with its own session crashes (timeout exc for instance). Are we sure both session are properly notified and rolled-back.

Now that the mongo migration is over, we may look at redesigning some aspects that were designed for session-less.

**Comment by benoit74** [comment]
@elfkuzco can you please investigate this? 

Question is what happens if some code runs "forever" in an API call, especially with the new stack? Is there something finally giving up and return an HTTP error code? What happens to the code running "forever"? What happens to the DB Lock we might have?

Pretty easy to simulate with an infinite loop in an API call which already grabbed a DB lock.

**Comment by elfkuzco** [comment]
Simulated with sleep of 10s while client made a request with timeout of 5s and API proceeded with request even after client timed out. Did some digging and it seems that in synchronous Python, you cannot stop the application automatically when the client disconnects. One way of preventing the query from continuing to run would be to enforce statement timeouts at the DB level https://stackoverflow.com/questions/24092463/psql-set-default-statement-timeout-as-a-user-in-postgres
There is also `transaction_timeout`s

**Comment by benoit74** [comment]
The problem here is not when client disconnects (this is normal) but rather that "something" was returning 499 and 504 errors while it looked like the server code was still running.

The stack now has significantly changed, but still:
- do we have something returning HTTP error to the client if the code takes too long (too long being probably at least 1 minute, maybe even more)
- if yes, what happens to the code running when this HTTP code is returned
- if no, could we implement something in FastAPI to stop code running for too long ? 

The idea is that even if we implement timeouts everywhere in the logic (in calls to S3, child HTTP requests to subservices, ...) , we might have a bug which cause code to run "forever" and if we do not have a very high level mechanism to stop this, then we might quickly get stuck (clients might retry the same HTTP call again, only causing more code running "indefinitely" in parallel). Since we usually start DB transaction very high in the API request handling, maybe it could do the trick. But I'm concerned about cases where we commit the transaction early, which would not be covered.

**Comment by elfkuzco** [comment]
> do we have something returning HTTP error to the client if the code takes too long (too long being probably at least 1 minute, maybe even more)

No

**Comment by elfkuzco** [comment]
> if no, could we implement something in FastAPI to stop code running for too long ?

given most of the code is synchronous. if it were asynchronous, we could do `asyncio.wait_for` on each endpoint. For synchronous, we would have to rely on some third party libs.

**Comment by benoit74** [comment]
Can't we do this at Uvicorn level then? Or should we put back something in front of it (Guvicorn? Nginx?)? How would it cooperate then?

**Comment by elfkuzco** [comment]
No. They don't seem to have a flag for it in the timeouts https://uvicorn.dev/settings/#timeouts

The timeout flags they have aren't particularly helpful

**Comment by elfkuzco** [comment]
Guvicorn has a timeout setting but it doesn't seem to apply when I use it with `uvicorn.workers.UvicornWorker` for FastAPI.

**Comment by benoit74** [comment]
OK, looking a bit more around on the Internet few things become clearer:
- FastAPI automatically runs it in a thread pool
- in Python, threads cannot be forcefully stopped by design (too much risk of leaving locks or whatever left-over)

There is hence no solution to this problem by design, except moving to an async design (ideally including async DB). Last attempt to move to async Python backend was not a success, so it has been ruled out for the time being.

I will hence close this issue because we have implemented sufficient safeguards in the codebase to not have long running requests in general.

We do not have problems of timeouts anymore ATM.

Should this become a problem, we will have to investigate and consider again the move to async API including DB.
